File: FPTrack-frontend/pages/registered_user/proposal.js
import Head from 'next/head'
import Image from 'next/image'
import styles from '../../styles/Home.module.css'
import Header from '../header'
import Link from 'next/link';
import { Multiselect } from 'multiselect-react-dropdown';
import { Component,useState } from 'react';
// title = "IoT Driven Smart Trains",
//         decription = "blah blah blah...",
//         domains = ['iot', 'machine_learning'],
//         supervisors = ['admin@admin.com', 'test@admin.com'],
//         leader = 'ben@gmail.com',
//         members = ['claire@gmail.com', 'derek@gmail.com'],
//         funding_type = 'internal',
//         funding_agency = 'SSN Trust',
//         pdf_document = pdf_doc_b64_string,
//         budget = 10000
export default function onProposal(props) {
	
    const [emails,setMail]=useState([]);
		const data = {
			domains: [{ name: 'machine_learning', id: 1 }, { name: 'web_development', id: 2 }, { name: 'iot', id: 3 }, { name: 'computer_vision', id: 4}],
		};

		const facultyPromise = new Promise((resolve, reject)=>{
				fetch("http://localhost:3000/api/user?role=faculty")
				.then(response => response.json())
				.then(jsondata => {
					var all_data = JSON.parse(
						JSON.stringify(jsondata)
					);
					// console.log(all_data);
					resolve(all_data);
				});
			}).then((data) => {
				var arr = [];
				var i=1;
				data.forEach((d) => {
					// console.log(this.data.supervisor);
          d.email.replace("@","jj");
					arr.push({'name' :"hi", 'id' : i})
					// console.log(d.email);
					i++;
				})
        
				console.log(arr);
			   //data.add({supervisor : arr});
        setMail(arr);
				// return arr;
			})
      console.log(data.domains);
      console.log(emails);
      
		
		const fileChangedHandler = (event) => {
			let file_size = event.target.files[0].size;
			console.log(file_size);
			if (file_size > 8451576) // approx 8MB
			{
				event.target.value = "";
				alert("File size should be less than 8Mb");
			}
    
			//do whatever operation you want to do here
		};

		return (
     
			<div className={styles.onProposal}>
				<Header></Header>

				<center>
					<form action="http://localhost:3000/api/proposal" className={styles.proposalForm} method="post">

						<legend>
							<h1> Project proposal </h1>
						</legend>

						<label>Title:<br /> <input type="text" name="title" /></label><br />
						<label>Description<br />
							<textarea name="description"></textarea> </label> <br />
						<label>Domains: <Multiselect //https://reactjsexample.com/react-multiselect-dropdown-with-search-and-various-options/
							style={{ // To change css for multiselect (Width,height,etc..)
								multiselectContainer: {
									width: 'fit-content'
								}
							}}
							options={data.domains} // Options to display in the dropdown
							// selectedValues={this.state.selectedValue} // Preselected value to persist in dropdown
							// onSelect={this.onSelect} // Function will trigger on select event
							// onRemove={this.onRemove} // Function will trigger on remove event
							displayValue="name" // Property name to display in the dropdown options
						/> </label><br />

						<label>Supervisors<br />
							{/* <textarea name ="supervisors"></textarea>  */}
							<Multiselect //https://reactjsexample.com/react-multiselect-dropdown-with-search-and-various-options/
                id ="supervisors"
								style={{
                  searchBox: { // To change search box element look
                      border: "none"
                    },
                  multiselectContainer: {
                      width: 'fit-content',
                      background: "white"
                  },
                  option: { // To change css for dropdown options
                      color: "black",
                    }
              }}
								options={emails==[]?None:emails}

								displayValue="supervisor"
							/>
						</label><br />
						<label>Leader<br /><input name="leader" type="email" /></label><br />
						<label>Members<br />
							<textarea name="members"></textarea> </label><br />
						<label >Funding_type<br />
							<select name="funding_type">
								<option value="internal">Internal</option>
								<option value="external">External</option>
							</select></label>
						<label>Funding_agency<br /><input name="funding_agency" type="text" /></label><br />
						<label>File upload<br /><input name="pdf_document" type="file" accept="application/pdf" onChange={fileChangedHandler} /></label><br />
						<label>Budget<br /><input name="budget" type="number" min="1" /></label><br />
						<input type="submit" />

					</form></center>

			</div>
		);
	
}

File: FPTrack-backend/api-tests/proposal.py
import requests
import base64
import os
import json

sample_pdf_path = os.path.join(os.getcwd(), *((os.path.pardir,)*1), 'check_files', 'agg_doc.pdf')
destn_pdf_path = os.path.join(os.getcwd(), *((os.path.pardir,)*1), 'check_files', 'retreive_doc.pdf')

def encode_pdf(filepath):
    pdf_doc_b64_bytes = base64.b64encode(open(filepath, 'rb').read())
    pdf_doc_b64_string = pdf_doc_b64_bytes.decode('utf-8')
    return pdf_doc_b64_string

def save_pdf(filedata, filepath):
    byte_data = decode_pdf(filedata)
    with open(filepath, 'wb') as f_out:
        f_out.write(byte_data)


def decode_pdf(byte_list):
    return base64.b64decode(''.join([chr(ascii) for ascii in byte_list]))


def post(debug=True):

    pdf_doc_b64_string = encode_pdf(sample_pdf_path)

    # req_body = dict(
    #     title = "IoT Driven Smart Trains",
    #     decription = "blah blah blah...",
    #     domains = ['iot', 'machine_learning'],
    #     supervisors = ['admin@admin.com', 'test@admin.com'],
    #     leader = 'ben@gmail.com',
    #     members = ['claire@gmail.com', 'derek@gmail.com'],
    #     funding_type = 'internal',
    #     funding_agency = 'SSN Trust',
    #     pdf_document = pdf_doc_b64_string,
    #     budget = 10000
    # )
    req_body = dict(
        title = "Number Plate Detection, Again",
        decription = "Blah blah blah...blah..",
        domains = ['machine_learning'],
        supervisors = ['sarabesh2001@gmail.com'],
        leader = 'ben@gmail.com',
        members = ['erin@gmail.com', 'greg@gmail.com'],
        funding_type = 'internal',
        funding_agency = 'SSN Trust',
        pdf_document = pdf_doc_b64_string,
        approved_on = "2020-12-12",
        rejected_on = "2020-12-14",
        budget = 50000
    )
    resp = requests.post(
        'http://localhost:3000/api/proposal',
        json = req_body
    )

    # print(resp.text)    
    print(resp.status_code)
    if not debug:
        resp = resp.json()[0]
        print(resp.keys())
        resp.pop('pdf_document')
        print(resp)


def get(id=None, debug=True):
    if id is None:
        resp = requests.get(
            'http://localhost:3000/api/proposal'
        )
    else:
        resp = requests.get(
            f'http://localhost:3000/api/proposal/{id}'
        )

    print(resp.text)
    print(resp.status_code)
    if not debug:
        if id is None:
            json_response = resp.json()[2]
        else:
            json_response = resp.json()
        print(json_response['_id'])
        print(json_response.keys())
        print(json_response['members'])
        resp = resp.json()[0]
        print(resp.keys())
        resp.pop('pdf_document')
        print(resp)
        save_pdf(json_response['pdf_document']['data'], destn_pdf_path)


def patch(debug=False):
    req_body = dict(
        id = '62912ac4f4ebb586b9b82e02',
        remarks = 'Needs more novelty'
    )
    resp = requests.patch(
        'http://localhost:3000/api/proposal/reject',
        json = req_body
    )

    print(resp.text)    
    print(resp.status_code)
    if not debug:
        print(resp.json())

# <<<<<<< HEAD
# post()
# get(debug=False)
get(id='6283b09950a9e277f0f8cd5a', debug=False)
# patch()